integration/auth_oidc_test.go

```diff
@@ -534,6 +539,211 @@ func TestOIDC024UserCreation(t *testing.T) {
+	)
+	assertNoErrHeadscaleEnv(t, err)
+
+	// Get all clients and verify they can connect
+	allClients, err := scenario.ListTailscaleClients()
+	assertNoErrListClients(t, err)
+
+	allIps, err := scenario.ListTailscaleClientsIPs()
+	assertNoErrListClientIPs(t, err)
+
+	err = scenario.WaitForTailscaleSync()
+	assertNoErrSync(t, err)
+
+	// Verify PKCE was used in authentication
+	headscale, err := scenario.Headscale()
+	assertNoErr(t, err)
+
+	var listUsers []v1.User
+	err = executeAndUnmarshal(headscale,
+		[]string{
+			"headscale",
+			"users",
+			"list",
+			"--output",
+			"json",
+		},
+		&listUsers,
+	)
+	assertNoErr(t, err)
+
+	allAddrs := lo.Map(allIps, func(x netip.Addr, index int) string {
+		return x.String()
+	})
+
+	success := pingAllHelper(t, allClients, allAddrs)
+	t.Logf("%d successful pings out of %d", success, len(allClients)*len(allIps))
+
+	// Verify all clients are connected and authenticated
+	for _, client := range allClients {
+		status, err := client.Status()
+		assertNoErr(t, err)
+		if status.BackendState != nodeStateRunning {
+			t.Errorf("client %s is not running: %s", client.Hostname(), status.BackendState)
+		}
+	}
+}
+
+type tamperVerifierTransport struct {
+	base http.RoundTripper
+}
+
+func (t *tamperVerifierTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	log.Printf("RoundTrip: %s %s", req.Method, req.URL.String())
+
+	// For POST requests, tamper with form data
+	if req.Method == http.MethodPost {
+		log.Printf("Processing POST request")
+		err := req.ParseForm()
+		if err != nil {
+			log.Printf("Error parsing form: %v", err)
+			return nil, err
+		}
+		if verifier := req.Form.Get("code_challenge"); verifier != "" {
+			log.Printf("Found POST verifier: %s", verifier)
+			// Tamper with the verifier
+			req.Form.Set("code_challenge", verifier+"_tampered")
+			log.Printf("Modified POST verifier to: %s", req.Form.Get("code_challenge"))
+			// Update request body with modified form
+			req.Body = io.NopCloser(strings.NewReader(req.Form.Encode()))
+			req.ContentLength = int64(len(req.Form.Encode()))
+		} else {
+			log.Printf("No code_challenge found in POST form data")
+		}
+	}
+
+	// For GET requests, tamper with URL query parameters
+	if req.Method == http.MethodGet {
+		log.Printf("Processing GET request")
+		q := req.URL.Query()
+		if verifier := q.Get("code_challenge"); verifier != "" {
+			log.Printf("Found GET verifier: %s", verifier)
+			q.Set("code_challenge", verifier+"_tampered")
+			req.URL.RawQuery = q.Encode()
+			log.Printf("Modified URL to: %s", req.URL.String())
+		} else {
+			log.Printf("No code_challenge found in GET query params")
+		}
+	}
+
+	// Forward the request with the tampered verifier
+	resp, err := t.base.RoundTrip(req)
+	if err != nil {
+		log.Printf("RoundTrip error: %v", err)
+
+		return nil, err
+	}
+	log.Printf("Response status: %s", resp.Status)
+
+	return resp, err
+}
+
+func TestOIDCAuthenticationWithPKCEVerifierTampering(t *testing.T) {
```

Comment:
I think the challenge with this test is that it doesnt verify that the test work, since the positive test uses the tailscale client, while the negative test does something completely different by running a http client. 

I am conflicted, I am not sure if this tests does what is needed, but not sure how we can achieve it with the client.

Reply:
Well, actually the negative test does something completely same as the positive test did, however the only difference is that I tried to modify the request from user to oidc server (with a custom http transport). The whole redirection logic is completely same:

1. User request login from tailscale
2. tailscale request headscale, headscale return machinekey `headscale.instance/register/mkey:xxxxxxxxxxx`
3. user request register endpoint, the endpoint redirect user to oidc login `oidc.instance/oidc/authorize?xxxxxx`
4. oidc server process request and redirect to headscale `headscale.instance/oidc/callback?xxxxxx`

What the http client did is that it modify the parameter `code_challenge` for the third step, so the login request from the headscale redirection is invalid. This is what we did to mimic MiTM. The oidc server will still proceed, but the response will not pass the code verifier stored inside headscale instance.

This is an actual log generated by the test:
```
2024/12/19 16:33:53 running tailscale up for user user1
2024/12/19 16:33:53 ts-unstable-dnh7da running tailscale up
2024/12/19 16:34:23 ts-unstable-dnh7da login url: https://172.22.0.5:8080/register/mkey:809f6ad985579a4ccabf7ee7880e20c7efb049c6f952dd174b6c93f617c50d1f
2024/12/19 16:34:23 ts-unstable-dnh7da logging in with url
2024/12/19 16:34:23 RoundTrip: GET https://172.22.0.5:8080/register/mkey:809f6ad985579a4ccabf7ee7880e20c7efb049c6f952dd174b6c93f617c50d1f
2024/12/19 16:34:23 Processing GET request
2024/12/19 16:34:23 No code_challenge found in GET query params
2024/12/19 16:34:23 Response status: 302 Found
2024/12/19 16:34:23 RoundTrip: GET http://172.22.0.3:34663/oidc/authorize?access_type=offline&client_id=superclient&code_challenge=6cg09A-VPrE-wgpCxUtfArY6ammErUF4vP3A2LXsm1M&code_challenge_method=S256&redirect_uri=https%3A%2F%2Fhs-oidcauthpkce-lrbl4l%3A8080%2Foidc%2Fcallback&response_type=code&scope=openid+profile+email&state=dcf4af7abc68da1876db3617094bf63b
2024/12/19 16:34:23 Processing GET request
2024/12/19 16:34:23 Found GET verifier: 6cg09A-VPrE-wgpCxUtfArY6ammErUF4vP3A2LXsm1M
2024/12/19 16:34:23 Modified URL to: http://172.22.0.3:34663/oidc/authorize?access_type=offline&client_id=superclient&code_challenge=6cg09A-VPrE-wgpCxUtfArY6ammErUF4vP3A2LXsm1M_tampered&code_challenge_method=S256&redirect_uri=https%3A%2F%2Fhs-oidcauthpkce-lrbl4l%3A8080%2Foidc%2Fcallback&response_type=code&scope=openid+profile+email&state=dcf4af7abc68da1876db3617094bf63b
2024/12/19 16:34:23 Response status: 302 Found
2024/12/19 16:34:23 RoundTrip: GET https://hs-oidcauthpkce-lrbl4l:8080/oidc/callback?code=bzi7BK5ZeYWCS3J7Es0PTIkKZ5riN4yP&state=dcf4af7abc68da1876db3617094bf63b
2024/12/19 16:34:23 Processing GET request
2024/12/19 16:34:23 No code_challenge found in GET query params
2024/12/19 16:34:23 Response status: 400 Bad Request
2024/12/19 16:34:23 ts-unstable-dnh7da response code of oidc login request was 400 Bad Request
2024/12/19 16:34:23 body: could not exchange code for token: oauth2: "invalid_grant" "Invalid code verifier. Code challenge did not match hashed code verifier."
2024/12/19 16:34:23 auth got error: status code not OK
```

Reply:
Unless I am really missing something, I think what bothers me is that it isnt the tailscale client that is doing the request, but a `httpClient.Do(req)`. I think the "correct" way to do this is to have the request go through a proxy and that could be passed as an option to hsic or tsic. where you can give it a func that will modify the requests.

That way we can have unmodified (and only one version of) `runTailscaleUp`, and have failing and passing nodes in the same test, some just get tampered with and some dont, compared to a dedicated test for each.

This is a lot of code that is very specific to one test, and a proxy in between headscale and tailscale for tampering would be quite useful. 

I dont want to put it on you to have to write that for this test tho, so I wonder if we will do the following: 

- Can you squash up the commits so there is two commits:
  - one with the code change, changelog and doc/config changes
  - one with the tests
- Make a PR with only the code change commit
- Make a draft PR with the tests
- Make an issue about adding a tampering proxy for headscale integration tests and tag the draft PR.

Then we can merge the code change without tests for now, and do it in a generic way so we dont have to maintain a lot of specific code.

Reply:
Ok, I see your point. by that way we can have a proxy which could also be useful for some other tests.